Speed up tensor.resize_(sizes) when tensor has correct size

zou3519 opened the pull request:
While using gbenchmark, I found `tensor.resize_({0})` would take 300ns
if tensor already has the correct size. This is important for
`at::empty({0})` perf because `at::empty` always calls `resize_`, which
in turn is a important for JIT perf: the fusion compiler creates empty
tensors and then `resize_`s them to computed sizes. Most of the 300ns is
due to DeviceGuard (200ns)

Summary of findings:
- `at::empty({0}, cuda)`: 851ns
- `empty_tensor.resize({0})`: 308ns
- `DeviceGuard(tensor)`: ctor + dtor: 200ns (Going to look into this
  next because it impacts `resize_` perf).
- vdispatch overhead (`tensor.resize_()` vs
  `at::native::resize__cuda(tensor)`): ~10ns

This PR rips out the TH `resize_` implementation and adds it to ATen
with the following modifications:
- DeviceGuard used only after the same-size check.
- Same-size check rewritten for simplicity. The new check doesn't
affect perf.
- empty_cpu / empty_cuda avoid the dispatch overhead to
tensor.resize_.

Timing with this PR:
- `at::empty({0}, cuda)`: 363ns
- `empty_tensor.resize_({0})`: 17ns

Future:
- Investigate `resize_(sizes)` slowness when `tensor.sizes() != sizes`
- Should tell resize_as_ to use the new resize_ implementation...
(because resize_as_ is in TH, it is calling the old TH resize_)

zou3519 committed:
Speed up tensor.resize_(sizes) when tensor has correct size

While using gbenchmark, I found `tensor.resize_({0})` would take 300ns
if tensor already has the correct size. This is important for
`at::empty({0})` perf because `at::empty` always calls `resize_`, which
in turn is a important for JIT perf: the fusion compiler creates empty
tensors and then `resize_`s them to computed sizes. Most of the 300ns is
due to DeviceGuard (200ns)

Summary of findings:
- `at::empty({0}, cuda)`: 851ns
- `empty_tensor.resize({0})`: 308ns
- `DeviceGuard(tensor)`: ctor + dtor: 200ns (Going to look into this
  next because it impacts `resize_` perf).
- vdispatch overhead (`tensor.resize_()` vs
  `at::native::resize__cuda(tensor)`): ~10ns

This PR rips out the TH `resize_` implementation and adds it to ATen
with the following modifications:
- DeviceGuard used only after the same-size check.
- Same-size check rewritten for simplicity. The new check doesn't
affect perf.
- empty_cpu / empty_cuda avoid the dispatch overhead to
tensor.resize_.

Timing with this PR:
- `at::empty({0}, cuda)`: 363ns
- `empty_tensor.resize_({0})`: 17ns

Future:
- Investigate `resize_(sizes)` slowness when `tensor.sizes() != sizes`
- Should tell resize_as_ to use the new resize_ implementation...
(because resize_as_ is in TH, it is calling the old TH resize_)

SsnL commented:
Nice. Not a blocker. But now you can call AT functions in TH I believe. :)

ezyang commented on aten/src/ATen/native/Resize.cpp:
```diff
@@ -0,0 +1,43 @@
+
+namespace at { namespace native {
+
+Tensor& resize__cpu(Tensor& self, IntList size) {
```
The return of the dastardly double underscore. (I laughed when I found out our codegen handles both naming conventions.)

zou3519 replied:
As in, our codegen handles `resize__cpu` and `resize_cpu_`?

ezyang replied:
yep

ezyang commented on aten/src/ATen/native/Resize.cpp:
```diff
@@ -0,0 +1,43 @@
+  }
+
+  for (int d = nDimension - 1; d >= 0; d--) {
+    self_->set_size(d, size[d]);
```
Oooh, please don't do it this way. Look at the implementation:

```
  virtual void set_size(int64_t dim, int64_t new_size) {
    sizes_.at(dim) = new_size;
    refresh_numel();
    refresh_contiguous();
  }
```

You're doing way more work than you need to. Use `set_sizes_and_strides` instead.

ezyang reviewed (changes requested):
Sorry, I know you're just trying to make this fast, but I'd like you to do a few more things:

1. Don't call `set_size` manually
2. Stack on top of https://github.com/pytorch/pytorch/pull/12845 so you can get contiguous strides allocation for free. (Or copy paste it in; it's an easy enough merge conflict to resolve)
3. Delete the old implementation, replacing it with a call to the `at::native::` function

Thanks!

zou3519 committed:
Cleanup resize_ code

zou3519 commented:
@ezyang I've addressed points (1) and (2). The conclusion for (3) is that there's no way to call aten native functions from TH at the moment but if that is not too hard to add I can punt this PR until that happens.

ezyang commented on aten/src/ATen/core/TensorImpl.h:
```diff
@@ -348,6 +348,40 @@ struct CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
+    sizes_ = new_size.vec();
+
+    // TODO(rzou): replace with update_to_contiguous_strides
+    // in #12845 when that lands.
```
Man, I hope we land that PR soon :)

ezyang reviewed (approved):
okey-dokey

zou3519 committed:
Cleanup: call new resize_ implementation from TH/THC

zou3519 commented:
@ezyang I figured out a compromise for (3): I've replaced the old TH implementation with the new implementation in this PR. I modified `TensorImpl::set_sizes_and_strides` to match the old `THTensor_(resizeNd)` semantics -- could you take a look and see if this is okay?

ezyang reviewed (approved):
OK. The code is still looks a bit copy-pastey, but I understand if you want to unblock.

zou3519 committed:
Cleanup

zou3519 committed:
Merge branch 'master' into resize

bddppq commented:
This PR has broken PyTorch ROCm tests:
https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-clang7-rocmdeb-ubuntu16.04-test/1261//console

cc @iotamudelta 

```
06:33:27 test_zeros_like (__main__.TestUncoalescedSparse) ... ok
06:33:28 
06:33:28 ======================================================================
06:33:28 ERROR: test_add_dense_sparse_mismatch (__main__.TestCudaSparse)
06:33:28 ----------------------------------------------------------------------
06:33:28 Traceback (most recent call last):
06:33:28   File "test_sparse.py", line 933, in test_add_dense_sparse_mismatch
06:33:28     test_shape([3, 4, 0], [1, 4], [4, 4, 4, 0], [3, 4, 4, 0])
06:33:28   File "test_sparse.py", line 923, in test_shape
06:33:28     x = torch.zeros(dense_size, dtype=self.value_dtype, device=self.device)
06:33:28 RuntimeError: cuda runtime error (1011) : hipErrorInvalidValue at /var/lib/jenkins/workspace/aten/src/THC/generic/THCTensorMath.cu:25
06:33:28 
06:33:28 ======================================================================
06:33:28 ERROR: test_add_dense_sparse_mismatch (__main__.TestCudaUncoalescedSparse)
06:33:28 ----------------------------------------------------------------------
06:33:28 Traceback (most recent call last):
06:33:28   File "test_sparse.py", line 933, in test_add_dense_sparse_mismatch
06:33:28     test_shape([3, 4, 0], [1, 4], [4, 4, 4, 0], [3, 4, 4, 0])
06:33:28   File "test_sparse.py", line 923, in test_shape
06:33:28     x = torch.zeros(dense_size, dtype=self.value_dtype, device=self.device)
06:33:28 RuntimeError: cuda runtime error (1011) : hipErrorInvalidValue at /var/lib/jenkins/workspace/aten/src/THC/generic/THCTensorMath.cu:25
06:33:28 
06:33:28 ======================================================================
06:33:28 ERROR: test_reduction_empty (test_torch.TestTorch)
06:33:28 ----------------------------------------------------------------------
06:33:28 Traceback (most recent call last):
06:33:28   File "/var/lib/jenkins/workspace/test/test_torch.py", line 1047, in test_reduction_empty
06:33:28     self.assertEqual(torch.empty((2, 0, 1), device=device), fn(x, dim=2, keepdim=True))
06:33:28   File "/var/lib/jenkins/workspace/test/test_torch.py", line 1024, in <lambda>
06:33:28     ('sum', lambda *args, **kwargs: torch.sum(*args, **kwargs), 0),
06:33:28 RuntimeError: cuda runtime error (1011) : hipErrorInvalidValue at /var/lib/jenkins/workspace/aten/src/THC/generic/THCTensorMath.cu:25
06:33:28 
06:33:28 ----------------------------------------------------------------------
```

jithunnair-amd commented:
This looks like more cases of "hipMemset with size 0"

iotamudelta commented:
so please skip for now, this will be fixed with `ROCm 1.9.2`
